docs/hugo/content/architecture-decision-records/ADR-2022-01-Reconciler-Extensions.md

```diff
@@ -0,0 +1,106 @@
+# Reconciler Extensions
+
+## Context
+
+We are discovering inconsistencies in the way different Azure Resource Providers behave.
+
+These inconsistencies are anticipated. Given the number of disparate teams independently implementing their providers and given the way the ARM guidance has changed over time, differences in behaviour are to be expected.
+
+For example, the Azure Redis Resource Provider can return a HTTP conflict (409) with a retry message indicating that it's not a fatal error but a transient one.
+
+Currently we handle this with special case behaviour in `error_classifier.go` (see [PR#2008](https://github.com/Azure/azure-service-operator/pull/2008) for details).
+
+However, directly changing our generic reconciler is a problem. The growing complexity makes this unsustainable, and there is a risk that changes made to support a new resource type will break an existing one.
+
+We need a way to customize its behaviour on a per-resource basis, allowing us to maintain consistent functionality within the operator.
+
+When a new version of a resource is introduced, we don't want to lose any existing customizations. This rules out hosting the extensions directly on the current hub type.
+
+It's quite likely that some extension points will need to import our generated resource types to allow per-version manipulation. This rules out referencing extensions from the generated resource types.
+
+We don't want extension points to reimplement functionality already present in the generic reconciler as this runs the risk of inconsistencies (especially if we change the generic reconciler itself), and introduces a significant overhead to implementation. Instead, the extension points will provide an opportunity to modify the information already held. An extension point that does nothing should have the same effect as one that hasn't been implemented.
+
+## Decision
+
+To provide a stable implementation location for extensions, we'll introduce a dedicated package alongside the existing versioned packages for each supported service. This package will provide extension support for all supported resource versions. For example, alongside the existing **batch** packages `v1alpha1api20210101`, and `v1alpha1api20210101storage`, we will generate `extensions`.
+
+Into the `extensions` package we will codegen a skeleton framework. For each resource there will be a separate extension type, suffixed with `Extensions`. For example, for the **compute** resources `Disk`, `VirtualMachine`, and `VirtualMachineScaleSet` we will create the extension types `DiskExtensions`, `VirtualMachineExtensions`, and `VirtualMachineScaleSetExtensions`.
```

Comment:
Is the advantage of pre-generating some things into `extensions` that we don't have to manually register things in the controller when an extension is added?

If we don't want to actually do the code generator work, we could instead just say that users must implement an `extensions` package (in that location) when they want to include an extension, and then register the extension in the `controller_resources.go` file (as opposed to the `controller_resources_gen.go` which is the generated one).

This has a few advantages in my head:
1. It avoids us having to generate templates, we could instead just write an example one manually, which is easier.
2. It avoids producing a bunch of code that does nothing - because an empty extension is the same as no extension at all. Much of the generated boilerplate will actually not be used in lots of packages where there is no need for an extension.
3. It avoids the "some of this package is generated and some of it is handcrafted" problem - it would mean that _everything_ in `extensions` is handcrafted. If we do generate templates we'll need to be careful about which files are generated and make sure to leave handcrafted stuff alone. That's fine but it places some restrictions on file layout.

Reply:
I want to make the overhead of writing an extenstion as low as possible, primarily so that it's easier for future maintainers to do the right thing than to start hacking around in the generic reconciler. Reducing the number of things that need to be known (or discovered) to correctly add an extension is key.

Generating the framework also allows us to constrain things. For example, by generating a host type, we can prevent it from having any state - which neatly sidesteps any issues caused by concurrent calls or extensions sharing state.

Doing less work is an attractive option though!

Leaving handcrafted stuff alone is easy - we already have the protocol of generating `_gen.go` files, and we only delete them if they have our header line in place. We'd generate a `VirtualMachineExtensions_gen.go` file, and any manually written code would go into `VirtualMachineExtensions.go`.

Reply:
> Leaving handcrafted stuff alone is easy

Yeah fair enough.

I guess what I am getting at is that we should look at what generating is giving us, and unless there's a lot of required boilerplate to get an extension working consider avoiding generating them. At least with my current understanding of the proposal I don't really see a ton of boilerplate, there's two parts:
1. Must create a type that implements the proposed extension interface
2. Must register that type with the controller

The actual implementation for item 1 has to be handcrafted (I think?), so code generation would just be making 2 easier - but I'm imagining 2 is basically a one-liner saying: `extensions.Add(&MySQLFlexibleServer{}, &MySQLFlexibleServerExtension{})`?